mage_ai/api/resources/PipelineRunResource.py

```diff
@@ -94,9 +98,38 @@ async def collection(self, query_arg, meta, user, **kwargs):
 
         repo_pipeline_schedule_ids = [s.id for s in PipelineSchedule.repo_query]
 
+        if status == PIPELINE_RUN_STATUS_LAST_RUN_FAILED:
+            """
+            The initial query below is used to get the last pipeline run retry
+            (or individual pipeline run if there are no retries) for each
+            grouping of pipeline runs with the same execution_date,
+            pipeline_uuid, and pipeline_schedule_id.
+            """
+            latest_pipeline_runs = PipelineRun.select(
+                PipelineRun.id,
+                func.row_number()
+                    .over(
+                        partition_by=(
+                            PipelineRun.execution_date,
+                            PipelineRun.pipeline_schedule_id,
+                            PipelineRun.pipeline_uuid,
+                        ),
+                        order_by=desc(PipelineRun.id))
+                    .label('row_number')
+            ).cte(name='latest_pipeline_runs')
+            query = (PipelineRun.select(
+                    PipelineRun,
+                )
+                .join(latest_pipeline_runs, and_(
+                    PipelineRun.id == latest_pipeline_runs.c.id,
+                    latest_pipeline_runs.c.row_number == 1,
+                ))
```

Comment:
what's the point of this join?

Reply:
@dy46 It joins the `latest_pipeline_runs` cte, which has the pipeline_run ids and row numbers to determine the latest pipeline run by `execution_date` (`pipeline_schedule_id` is used to sort the runs because there can be runs with the exact same `created_at` dates, which I tried initially), with the regular PipelineRun table and also filters out any rows that do NOT have row_number of `1`. The rows with row_number of 1 are the latest pipeline run.

Maybe a little confusing combining the filter in the join. Another way of writing the query, would this be clearer?
```
query = (PipelineRun.select(
                    PipelineRun,
                )
                # The `and_` is needed even though there's only 1 condition or the query doesn't work properly.
                .join(latest_pipeline_runs, and_(PipelineRun.id == latest_pipeline_runs.c.id))
                .filter(latest_pipeline_runs.c.row_number == 1)
            )
```

Reply:
ah yeah i get it - i forgot you have to get the actual pipeline run object. i think the original way you did it is ok